File: app/src/main/java/com/example/tanawatk/mvvmandroid/di/module/AppModule.java
package com.example.tanawatk.mvvmandroid.di.module;

import android.app.Application;
import android.content.Context;

import com.example.tanawatk.mvvmandroid.di.qualifier.ApplicationContext;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {

    Application application;

    public AppModule(Application application) {
        this.application = application;
    }

    @Singleton
    @Provides
    @ApplicationContext
    Context provideApplicationContext() {
        return application.getApplicationContext();
    }

    @Singleton
    @Provides
    Application provideApplication(){
        return application;
    }
}

File: app/src/main/java/com/example/tanawatk/mvvmandroid/di/module/AppModule.kt
package com.example.tanawatk.mvvmandroid.di.module

import android.app.Application
import android.content.Context
import com.example.tanawatk.mvvmandroid.di.qualifier.ApplicationContext
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class AppModule(private val application: Application) {

    @Singleton @Provides @ApplicationContext
    fun provideApplicationContext(): Context = application.applicationContext

    @Singleton @Provides
    fun provideApplication(): Application = application
}